src/server/twemcache-rs/src/worker/single.rs

```diff
@@ -0,0 +1,272 @@
+                            self.handle_error(token);
+                        }
+                    }
+                }
+                Err(_) => {
+                    error!("Error registering new socket");
+                }
+            };
+        }
+    }
+
+    fn handle_event(&mut self, event: &Event) {
+        let token = event.token();
+
+        // event for existing session
+        trace!("got event for session: {}", token.0);
+
+        // handle error events first
+        if event.is_error() {
+            increment_counter!(&Stat::WorkerEventError);
+            self.handle_error(token);
+        }
+
+        // handle write events before read events to reduce write buffer
+        // growth if there is also a readable event
+        if event.is_writable() {
+            increment_counter!(&Stat::WorkerEventWrite);
+            self.do_write(token);
+        }
+
+        // read events are handled last
+        if event.is_readable() {
+            increment_counter!(&Stat::WorkerEventRead);
+            let _ = self.do_read(token);
+
+            #[cfg(feature = "heap_dump")]
+            {
+                ops += 1;
+                if ops >= 1_000_000 {
+                    let dump = self.data.dump();
+                    let serialized = serde_json::to_string(&dump).unwrap();
+                    let mut file = std::fs::File::create(&format!("dump_{}.raw", seq)).unwrap();
+                    let _ = file.write_all(serialized.as_bytes());
+                    seq += 1;
+                    ops = 0;
+                }
+            }
+        }
+
+        if let Some(session) = self.sessions.get_mut(token.0) {
+            trace!(
+                "{} bytes pending in rx buffer for session: {}",
+                session.read_pending(),
+                token.0
+            );
+            trace!(
+                "{} bytes pending in tx buffer for session: {}",
+                session.write_pending(),
+                token.0
+            )
+        }
+    }
+
+    pub fn message_sender(&self) -> Sender<Message> {
+        self.message_sender.clone()
+    }
+
+    pub fn session_sender(&self) -> Sender<Session> {
+        self.session_sender.clone()
+    }
+}
+
+impl EventLoop for SingleWorker {
+    fn get_mut_session(&mut self, token: Token) -> Option<&mut Session> {
+        self.sessions.get_mut(token.0)
+    }
+
+    fn handle_data(&mut self, token: Token) -> Result<(), ()> {
+        trace!("handling request for session: {}", token.0);
+        if let Some(session) = self.sessions.get_mut(token.0) {
+            loop {
+                // TODO(bmartin): buffer should allow us to check remaining
+                // write capacity.
+                if session.write_pending() > MIN_BUFFER_SIZE {
+                    // if the write buffer is over-full, skip processing
+                    break;
+                }
+                match MemcacheParser::parse(&mut session.read_buffer) {
```

Comment:
Similar comment tot he multi-worker. I think we can expect a generic `RequestProcessor` to have a few traits that are relevant to eventloop or IPC, for single, that'd just be "parse" (buffer->request), "execute" (request->response), and "compose" (response->buffer). For multi, we additionally need send/recv communication with storage.

Reply:
Maybe this is a naming thing - or maybe I went with the wrong abstraction.

The `RequestProcessor` is serving as a common location for defining how to take a parsed request and some storage and write a response back to the session buffer. This is really just allowing us to share code between the single worker and the storage thread. It knows the semantics of the underlying storage - for example how to combine primitive storage operations (get, set) into add/replace - and that the optional data is variable length in segcache but used as a u32 flag for memcache protocol.